packages/core/src/modules/discover-features/DiscoverFeaturesModule.ts

```diff
@@ -82,18 +94,44 @@ export class DiscoverFeaturesModule {
     return isProtocolSupported
   }
 
-  public async queryFeatures(connectionId: string, options: { query: string; comment?: string }) {
+  /**
+   * Send a query to an existing connection for discovering supported features. Depending on options, it will do a
+   * simple query through Discover Features V1 protocol (Aries RFC 0031) or multiple feature query by Discover
+   * Features V2 (Aries RFC 0557).
+   *
+   * @param connectionId connection to query features from
+   * @param options query string for simple protocol support (using Discover Features V1) or set of queries to ask
+   * multiple features (using Discover Features V2 protocol). Optional comment string only used for simple queries.
+   */
+  public async queryFeatures(
+    connectionId: string,
+    options: { query: string | FeatureQueryOptions[]; comment?: string }
+  ) {
     const connection = await this.connectionService.getById(connectionId)
 
-    const queryMessage = await this.discoverFeaturesService.createQuery(options)
+    const queryMessage =
+      typeof options.query === 'string'
+        ? await this.discoverFeaturesService.createQuery({
+            query: options.query,
+            comment: options.comment,
+          })
+        : await this.discoverFeaturesV2Service.createQueries({ queries: options.query })
```

Comment:
Not too sure on swithcing the protocol version based on the query format.

What if we model it more like we do with the credential/proofs modules where the input is the same but you pass the protocol version? This will also help if a v3 is added that is the same but just uses didcomm v2. If you pass a query that is too advanced for v1 it will throw an error (like you can't issue jsonld with v1 issue credential).

Or we trigger the options type based on the protocol version. So it is still a different query type, but you MUST provide the protocolVersion

Reply:
I do also think that's the best way to proceed. However, I don't want to break current API yet so I guess I have to get some classes of TypeScript Magic to achieve that :smile:  Maybe we can add an optional parameter with protocol version that defaults to V1 and switches the expected parameters, which is more or less what you suggest if I'm not wrong.